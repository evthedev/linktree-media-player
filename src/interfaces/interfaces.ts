import { DefaultColors, MusicProvider, MediaType } from '../common/constants';

export interface IRoute {
	path: string;
	exact: boolean;
	sidebar: string;
	icon: React.FC,
	main: any
}

export interface IMediaOwnProps {
	type: MediaType.Classic | MediaType.Music | MediaType.Show
}

export interface IMediaBaseProps {
	backgroundColor: string;
	fontColor: string;
}

export interface IClassic extends IMediaBaseProps {
	id: number;
	link: string
}

export interface IShow extends IMediaBaseProps {
	id: number;
	date: string;
	location: string;
	url: string;
	ticketsAvailable: number
}

export interface IMusic extends IMediaBaseProps {
	id: number;
	services: IService[]
}
export interface IService {
	id: number;
	providerName: MusicProvider;
	trackUrl: string
}

// export interface IAddress {
// 	id: number;
// 	quarantine_address: string;
// 	quarantine_address_lat: number | string;
// 	quarantine_address_long: number | string;
// }
// export interface ISanitisedAddress {
// 	id: number;
// 	quarantineAddress: string;
// 	quarantineAddressLat: number;
// 	quarantineAddressLong: number;
// }

// export interface ISanitisedUser {
// 	date: string;
// 	arrivedTravellers: number;
// 	totalUsersProvidedConsent: number;
// 	travellersArrivedAndProvidedConsent: number;
// 	travellersArrivedAndCompletedSetup: number;
// }

// export interface IPill {
// 	backgroundColor: DefaultColors
// }

export interface IState {
	links: IClassic[],
	shows: IShow[],
	music: IMusic[]
}

// export interface IMapProps {
// 	addresses: ISanitisedAddress[]
// }