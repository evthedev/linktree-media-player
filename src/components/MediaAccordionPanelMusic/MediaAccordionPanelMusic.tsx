import React, { Fragment, useEffect, useState } from 'react';
import { MediaStyledPanel } from '../../common/styles';
import { IMusic, IShow, IService } from '../../interfaces/interfaces';

import iconRightArrow from '../../assets/images/icon-right-arrow@2x.png';

import { players } from '../../config/players';
import { defaultService } from '../../common/defaults';

export const MediaAccordionPanelMusic: React.FC<IMusic | IShow> = (props): JSX.Element => {

	const { services } = props as IMusic;

	const [ activeService, setActiveService ] = useState<IService>(defaultService);
	
	const playerComponent = players.find(({musicProvider}) => musicProvider === activeService.providerName)?.component;

	useEffect(() => {
		// set first item of services as the embedded player
		setActiveService(services[0]);
	}, [services]);

	const handlePanelClick = (servicesItem: IService) => {
		setActiveService(servicesItem);
	}
	
	return (
		<Fragment>			
			{ playerComponent && playerComponent(activeService.trackUrl) }
			{ services && services.map(servicesItem =>
				<MediaStyledPanel className="media-panel__panel" key={servicesItem.id} onClick={() => handlePanelClick(servicesItem)}>
					<img className="media-panel__icon" src={ players.find(({musicProvider}) => musicProvider === servicesItem.providerName)?.logo } alt={servicesItem.providerName} />
					<h6 className="media-panel__title">{ servicesItem.providerName }</h6>
					<img className="media-panel__arrow" src={iconRightArrow} alt="icon-right-arrow"/>
				</MediaStyledPanel>
			)}			
		</Fragment>
	)
}
