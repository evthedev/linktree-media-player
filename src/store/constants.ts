import { IClassicLink, IShow, IMusic, IMediaBaseProps } from '../interfaces/interfaces';

export const SET_MEDIA_BUTTON_STYLES = 'SET_MEDIA_BUTTON_STYLES';
export const SET_MEDIA_DATA = 'SET_MEDIA_DATA';

export interface SetMediaDataAction {
	type: typeof SET_MEDIA_DATA;
	payload: IMusic[] | IShow[] | IClassicLink[]
}

export interface setMediaButtonStylesAction {
	type: typeof SET_MEDIA_BUTTON_STYLES;
	payload: IMediaBaseProps
}

export type actionTypes = SetMediaDataAction | setMediaButtonStylesAction