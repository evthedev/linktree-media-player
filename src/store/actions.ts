import { Dispatch } from 'react';
import { IClassicLink, IMediaBaseProps, IMusic, IShow } from '../interfaces/interfaces';
import { actionTypes, SET_MEDIA_BUTTON_STYLES, SET_MEDIA_DATA } from "./constants";
import { MediaType } from '../common/constants';
import { commandHandlers } from '../config/commandHandlers';

export const setMediaButtonStyles = (mediaButtonStyles: IMediaBaseProps): actionTypes => ({
	type: SET_MEDIA_BUTTON_STYLES,
	payload: mediaButtonStyles
})

export const setMediaData = (mediaData: IClassicLink[] | IShow[] | IMusic[]): actionTypes => ({
	type: SET_MEDIA_DATA,
	payload: mediaData
})

export const clearData = () => (
	(dispatch: Dispatch<actionTypes>) => {
		dispatch(setMediaData([]));
	}
)

export const fetchFakeMedia = (type: MediaType) => (
	async (dispatch: Dispatch<actionTypes>) => {
		// Get actionHandler for the specific media type
		
		// Get data for the specific media type
		const data = commandHandlers.filter(commandHandler => commandHandler.mediaType === type)[0].data;
		
		// Use a delay to simulate a delay when fetching from 'server'
		setTimeout(() => {
			dispatch(setMediaButtonStyles({
				backgroundColor: data.backgroundColor,
				fontColor: data.fontColor
			}))
			dispatch(setMediaData(data.results))
		}, 200)

	}
)
