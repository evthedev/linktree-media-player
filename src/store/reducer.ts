import { defaultState } from '../common/defaults';
import { IState } from '../interfaces/interfaces';
import { actionTypes, SET_MEDIA_BUTTON_STYLES, SET_MEDIA_DATA } from './constants';

export const reducer = (state: IState = defaultState, action: actionTypes): IState => {
	switch (action.type) {
		case SET_MEDIA_DATA:
			return {
				...state,
				mediaData: action.payload
			}	
		case SET_MEDIA_BUTTON_STYLES:
			return {
				...state,
				backgroundColor: action.payload.backgroundColor,
				fontColor: action.payload.fontColor
			}	
		default:
			return state		
	}
}
