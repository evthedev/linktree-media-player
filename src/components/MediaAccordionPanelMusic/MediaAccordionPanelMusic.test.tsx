import React from 'react';
import Enzyme, { mount } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import { IMusic, IState } from '../../interfaces/interfaces';
import { reducer } from '../../store/reducer';

import { MediaAccordionPanelMusic } from './MediaAccordionPanelMusic';

Enzyme.configure({ adapter: new EnzymeAdapter()});

describe('<MediaAccordionPanelMusic /> component', () => {
	const mockInitialState: IState = {
		backgroundColor: '',
		fontColor: '',
		mediaData: [],
	}

	const testMusicData: IMusic = {
		id: 1,
		title: "Test song 1",
		services: [
			{
				id: 1,
				providerName: "Spotify",
				trackUrl: "https://open.spotify.com/embed/track/sfasgasouytt74"
			},
		]
	}

	const getWrapper = (mockStore = createStore(reducer, mockInitialState)) => mount(
		<Provider store={mockStore}>
			<MediaAccordionPanelMusic {...testMusicData} />
		</Provider>
	)
	it('should mount correctly with menu items', () => {
		const wrapper = getWrapper();
		expect(wrapper.isEmptyRender()).toBe(false);
		expect(wrapper).toMatchSnapshot();
	})
})