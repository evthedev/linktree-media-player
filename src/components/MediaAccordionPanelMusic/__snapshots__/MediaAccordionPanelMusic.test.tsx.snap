// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaAccordionPanelMusic /> component should mount correctly with menu items 1`] = `ReactWrapper {}`;
